sop-extras/src/Data/SOP/Index.hs

```diff
@@ -36,29 +39,42 @@ module Data.SOP.Index (
   ) where
 
 import           Data.Coerce
-import           Data.Kind (Type)
 import           Data.Proxy
 import           Data.SOP.BasicFunctors
 import           Data.SOP.Constraint
 import           Data.SOP.Dict (Dict (..))
 import           Data.SOP.Sing
 import           Data.SOP.Strict
+import           Data.Type.Equality ((:~:) (..))
 import           Data.Word
 
-type Index :: [k] -> k -> Type
-data Index xs x where
-  IZ ::               Index (x ': xs) x
-  IS :: Index xs x -> Index (y ': xs) x
+newtype Index xs x = Index { getIndex :: NS ((:~:) x) xs }
+
+pattern IZ ::
+     ()
+  => xs ~ (x ': xs1)
+  => Index xs x
+pattern IZ = Index (Z Refl)
+
+pattern IS ::
+     ()
+  => xs ~ (x' ': xs')
+  => Index xs' x
+  -> Index xs x
+pattern IS z <- Index (S (Index -> z)) where
+  IS (Index z) = Index (S z)
+
+{-# COMPLETE IZ, IS #-}
```

Comment:
Hmm, while this approach is indeed makes the payload strict, GHC's pattern match coverage checker does not seem to detect that when using the pattern synonyms:
```haskell
test :: Index '[()] () -> ()
test IZ     = ()
test (IS a) = case a of {}
```
still requires the `IS` case, even though it is now definitely unreachable. :thinking: 

In contrast, keeping the previous definition of `Index` and making the `IS` payload strict should allow simplifying `test`.

Not a disaster, but something to be aware of.

Reply:
This is unfortunate. Do we want to proceed or shall we keep the previous definition and make it strict? I think it is fine this way because we can make use of `EmptyCase` to signal that match is inhabited.